### PR TITLE
rpc: Version the RPC APIs

### LINKED ISSUE #75: rpc: Version the RPC APIs
At present, we support a single version of RPC API on a node. I propose we version our RPC API such that we can roll out substantial API changes in a non-breaking way.

Specifically, our current RPC API (JSON, HTTP, and Websocket) is available at the root of the node, e.g. `/status`, `/block`, etc.
We need to 

- [ ] Make our current RPC API is _also_ available under a `/v1` sub-path, e.g. `/v1/status`, `/v1/block`, etc.
    - [ ] JSON
    - [ ] Websocket
    - [ ] HTTP
- [ ] Publish the extended API

With this change we can easily roll out new RPC API versions with "breaking" changes as compared to previous versions, e.g. under a `/v2`, `/v3`, etc. path

Eventually we would be able to deprecate the APIs available at the root, then we'd deprecate the `/v1` API, etc. but this process would be progressive and consultative with integrators.

### PR DESCRIPTION
**Pull request by lasarojc**
Closes #75

---

#### PR checklist

- [ ] Tests written/updated
- [ ] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [ ] Updated relevant documentation (`docs/` or `spec/`) and code comments

### CODE REVIEW

**Comment by lasarojc** [comment]
The versioned methods will have a vx/ prefix, as in v1/block.
v1 is the current version and we are just now making it explicit.
v1 will continue to be accessible without the prefix, while this version is supported.

At some point we will need a method to get the list of supported versions, but we will wait until other changes to the API are needed, before introducing such a method.

**Comment by lasarojc** [comment]
In this PR I differentiate between v0.3, used in comet up to v0.38, and v1, used in the following version.
Although they have exactly the same functionality, they are different API since they have different endpoints (in v1, the methods are prefixed with v1.

I considered simply aliasing the methods using `$ref`, but that would imply in either removing `operationID` from the methods definitions, or in duplicating all the endpoints, which seemed wrong. With the differentiation v0.3 x v1, anyone using tooling on the openapi definition will get the expected behavior (e.g., when generating client code).

**Comment by lasarojc** [comment]
> At some point we will need a method to get the list of supported versions, but we will wait until other changes to the API are needed, before introducing such a method.

Actually, one can get the list of supported methods by querying the root rpm endpoint. I.e., e.g.

```sh
curl http://127.0.0.1:26657
```
returns
```html
<html><body><br>Available endpoints:<br><br>Endpoints that require arguments:<br><a href="//127.0.0.1:26657/abci_info?">//127.0.0.1:26657/abci_info?</a></br><a href="//127.0.0.1:26657/abci_query?path=_&data=_&height=_&prove=_">//127.0.0.1:26657/abci_query?path=_&data=_&height=_&prove=_</a></br><a href="//127.0.0.1:26657/block?height=_">//127.0.0.1:26657/block?height=_</a></br><a href="//127.0.0.1:26657/block_by_hash?hash=_">//127.0.0.1:26657/block_by_hash?hash=_</a></br><a href="//127.0.0.1:26657/block_results?height=_">//127.0.0.1:26657/block_results?height=_</a></br><a href="//127.0.0.1:26657/block_search?query=_&page=_&per_page=_&order_by=_">//127.0.0.1:26657/block_search?query=_&page=_&per_page=_&order_by=_</a></br><a href="//127.0.0.1:26657/blockchain?minHeight=_&maxHeight=_">//127.0.0.1:26657/blockchain?minHeight=_&maxHeight=_</a></br><a href="//127.0.0.1:26657/broadcast_evidence?evidence=_">//127.0.0.1:26657/broadcast_evidence?evidence=_</a></br><a href="//127.0.0.1:26657/broadcast_tx_async?tx=_">//127.0.0.1:26657/broadcast_tx_async?tx=_</a></br><a href="//127.0.0.1:26657/broadcast_tx_commit?tx=_">//127.0.0.1:26657/broadcast_tx_commit?tx=_</a></br><a href="//127.0.0.1:26657/broadcast_tx_sync?tx=_">//127.0.0.1:26657/broadcast_tx_sync?tx=_</a></br><a href="//127.0.0.1:26657/check_tx?tx=_">//127.0.0.1:26657/check_tx?tx=_</a></br><a href="//127.0.0.1:26657/commit?height=_">//127.0.0.1:26657/commit?height=_</a></br><a href="//127.0.0.1:26657/consensus_params?height=_">//127.0.0.1:26657/consensus_params?height=_</a></br><a href="//127.0.0.1:26657/consensus_state?">//127.0.0.1:26657/consensus_state?</a></br><a href="//127.0.0.1:26657/dump_consensus_state?">//127.0.0.1:26657/dump_consensus_state?</a></br><a href="//127.0.0.1:26657/genesis?">//127.0.0.1:26657/genesis?</a></br><a href="//127.0.0.1:26657/genesis_chunked?chunk=_">//127.0.0.1:26657/genesis_chunked?chunk=_</a></br><a href="//127.0.0.1:26657/header?height=_">//127.0.0.1:26657/header?height=_</a></br><a href="//127.0.0.1:26657/header_by_hash?hash=_">//127.0.0.1:26657/header_by_hash?hash=_</a></br><a href="//127.0.0.1:26657/health?">//127.0.0.1:26657/health?</a></br><a href="//127.0.0.1:26657/net_info?">//127.0.0.1:26657/net_info?</a></br><a href="//127.0.0.1:26657/num_unconfirmed_txs?">//127.0.0.1:26657/num_unconfirmed_txs?</a></br><a href="//127.0.0.1:26657/status?">//127.0.0.1:26657/status?</a></br><a href="//127.0.0.1:26657/subscribe?query=_">//127.0.0.1:26657/subscribe?query=_</a></br><a href="//127.0.0.1:26657/tx?hash=_&prove=_">//127.0.0.1:26657/tx?hash=_&prove=_</a></br><a href="//127.0.0.1:26657/tx_search?query=_&prove=_&page=_&per_page=_&order_by=_">//127.0.0.1:26657/tx_search?query=_&prove=_&page=_&per_page=_&order_by=_</a></br><a href="//127.0.0.1:26657/unconfirmed_txs?limit=_">//127.0.0.1:26657/unconfirmed_txs?limit=_</a></br><a href="//127.0.0.1:26657/unsubscribe?query=_">//127.0.0.1:26657/unsubscribe?query=_</a></br><a href="//127.0.0.1:26657/unsubscribe_all?">//127.0.0.1:26657/unsubscribe_all?</a></br><a href="//127.0.0.1:26657/v1/abci_info?">//127.0.0.1:26657/v1/abci_info?</a></br><a href="//127.0.0.1:26657/v1/abci_query?path=_&data=_&height=_&prove=_">//127.0.0.1:26657/v1/abci_query?path=_&data=_&height=_&prove=_</a></br><a href="//127.0.0.1:26657/v1/block?height=_">//127.0.0.1:26657/v1/block?height=_</a></br><a href="//127.0.0.1:26657/v1/block_by_hash?hash=_">//127.0.0.1:26657/v1/block_by_hash?hash=_</a></br><a href="//127.0.0.1:26657/v1/block_results?height=_">//127.0.0.1:26657/v1/block_results?height=_</a></br><a href="//127.0.0.1:26657/v1/block_search?query=_&page=_&per_page=_&order_by=_">//127.0.0.1:26657/v1/block_search?query=_&page=_&per_page=_&order_by=_</a></br><a href="//127.0.0.1:26657/v1/blockchain?minHeight=_&maxHeight=_">//127.0.0.1:26657/v1/blockchain?minHeight=_&maxHeight=_</a></br><a href="//127.0.0.1:26657/v1/broadcast_evidence?evidence=_">//127.0.0.1:26657/v1/broadcast_evidence?evidence=_</a></br><a href="//127.0.0.1:26657/v1/broadcast_tx_async?tx=_">//127.0.0.1:26657/v1/broadcast_tx_async?tx=_</a></br><a href="//127.0.0.1:26657/v1/broadcast_tx_commit?tx=_">//127.0.0.1:26657/v1/broadcast_tx_commit?tx=_</a></br><a href="//127.0.0.1:26657/v1/broadcast_tx_sync?tx=_">//127.0.0.1:26657/v1/broadcast_tx_sync?tx=_</a></br><a href="//127.0.0.1:26657/v1/check_tx?tx=_">//127.0.0.1:26657/v1/check_tx?tx=_</a></br><a href="//127.0.0.1:26657/v1/commit?height=_">//127.0.0.1:26657/v1/commit?height=_</a></br><a href="//127.0.0.1:26657/v1/consensus_params?height=_">//127.0.0.1:26657/v1/consensus_params?height=_</a></br><a href="//127.0.0.1:26657/v1/consensus_state?">//127.0.0.1:26657/v1/consensus_state?</a></br><a href="//127.0.0.1:26657/v1/dump_consensus_state?">//127.0.0.1:26657/v1/dump_consensus_state?</a></br><a href="//127.0.0.1:26657/v1/genesis?">//127.0.0.1:26657/v1/genesis?</a></br><a href="//127.0.0.1:26657/v1/genesis_chunked?chunk=_">//127.0.0.1:26657/v1/genesis_chunked?chunk=_</a></br><a href="//127.0.0.1:26657/v1/header?height=_">//127.0.0.1:26657/v1/header?height=_</a></br><a href="//127.0.0.1:26657/v1/header_by_hash?hash=_">//127.0.0.1:26657/v1/header_by_hash?hash=_</a></br><a href="//127.0.0.1:26657/v1/health?">//127.0.0.1:26657/v1/health?</a></br><a href="//127.0.0.1:26657/v1/net_info?">//127.0.0.1:26657/v1/net_info?</a></br><a href="//127.0.0.1:26657/v1/num_unconfirmed_txs?">//127.0.0.1:26657/v1/num_unconfirmed_txs?</a></br><a href="//127.0.0.1:26657/v1/status?">//127.0.0.1:26657/v1/status?</a></br><a href="//127.0.0.1:26657/v1/subscribe?query=_">//127.0.0.1:26657/v1/subscribe?query=_</a></br><a href="//127.0.0.1:26657/v1/tx?hash=_&prove=_">//127.0.0.1:26657/v1/tx?hash=_&prove=_</a></br><a href="//127.0.0.1:26657/v1/tx_search?query=_&prove=_&page=_&per_page=_&order_by=_">//127.0.0.1:26657/v1/tx_search?query=_&prove=_&page=_&per_page=_&order_by=_</a></br><a href="//127.0.0.1:26657/v1/unconfirmed_txs?limit=_">//127.0.0.1:26657/v1/unconfirmed_txs?limit=_</a></br><a href="//127.0.0.1:26657/v1/unsubscribe?query=_">//127.0.0.1:26657/v1/unsubscribe?query=_</a></br><a href="//127.0.0.1:26657/v1/unsubscribe_all?">//127.0.0.1:26657/v1/unsubscribe_all?</a></br><a href="//127.0.0.1:26657/v1/validators?height=_&page=_&per_page=_">//127.0.0.1:26657/v1/validators?height=_&page=_&per_page=_</a></br><a href="//127.0.0.1:26657/validators?height=_&page=_&per_page=_">//127.0.0.1:26657/validators?height=_&page=_&per_page=_</a></br></body></html>%
```

File: .changelog/unreleased/improvements/1412-rpc-versioning.md [hunk 1]
```diff
@@ -0,0 +1,2 @@
+- `[rpc]` The RPC API version is now reflected on the endpoint.
```
Comment (thanethomson) [inline .changelog/unreleased/improvements/1412-rpc-versioning.md:1]:
```suggestion
- `[rpc]` The RPC API is now versioned, with all existing endpoints accessible
  via `/v1/*` as well as `/*`
```

File: UPGRADING.md [hunk 2]
```diff
@@ -41,6 +43,13 @@ The minimum Go version has been bumped to [v1.21][go121].
 * Removed the `replay` and `replay-console` subcommands
   ([\#1170](https://github.com/cometbft/cometbft/pull/1170))
 
+### RPC API
+
+* The RPC API version is now reflected on the endpoint.
```
Comment (thanethomson) [inline UPGRADING.md:48]:
```suggestion
* The RPC API is now versioned.
```

File: UPGRADING.md [hunk 3]
```diff
@@ -41,6 +43,13 @@ The minimum Go version has been bumped to [v1.21][go121].
...
+  Although invoking methods without specifying the version is still supported for now,
+  support will be dropped in future releases and users are urged to use the new format.
```
Comment (thanethomson) [inline UPGRADING.md:50]:
```suggestion
  support will be dropped in future releases and users are urged to use the versioned
  approach.
```

**Review by thanethomson — COMMENTED** [review]
This looks good so far!

@lasarojc would you mind adapting all of the E2E tests to use the `/v1/` endpoints instead? Otherwise we're not _really_ testing this change. Just by changing this part of the E2E code, we should be able to test this: https://github.com/cometbft/cometbft/blob/6f2590df767be4c1824f1cc4070a647c417e6e75/test/e2e/pkg/testnet.go#L555-L557

File: rpc/jsonrpc/server/http_json_handler.go [hunk 4]
```diff
@@ -132,7 +132,7 @@ func handleInvalidJSONRPCPaths(next http.HandlerFunc) http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
 		// Since the pattern "/" matches all paths not matched by other registered patterns,
 		//  we check whether the path is indeed "/", otherwise return a 404 error
```
Comment (mzabaluev) [inline rpc/jsonrpc/server/http_json_handler.go:134]: The comment needs to be updated as well, I think.

File: consensus/reactor_test.go [hunk 5]
```diff
@@ -432,8 +432,8 @@ func TestReactorRecordsVotesAndBlockParts(t *testing.T) {
 	// Get peer state
 	ps := peer.Get(types.PeerStateKey).(*PeerState)
 
-	assert.Equal(t, true, ps.VotesSent() > 0, "number of votes sent should have increased")
-	assert.Equal(t, true, ps.BlockPartsSent() > 0, "number of votes sent should have increased")
+	assert.Greater(t, ps.VotesSent(), 0, "number of votes sent should have increased")
+	assert.Greater(t, ps.BlockPartsSent(), 0, "number of votes sent should have increased")
```
Comment (mzabaluev) [inline consensus/reactor_test.go:436]: This looks unrelated.

File: test/e2e/docker/Dockerfile.debug [hunk 6]
```diff
@@ -20,6 +20,7 @@ RUN go mod download
 COPY . .
 RUN echo $COMETBFT_BUILD_OPTION && make build && cp build/cometbft /usr/bin/cometbft
 COPY test/e2e/docker/entrypoint* /usr/bin/
+COPY test/e2e/docker/entrypoint-builtin-delve /usr/bin/entrypoint-builtin
```
Comment (mzabaluev) [inline test/e2e/docker/Dockerfile.debug:23]: These two lines mean there will be an exact copy of `entrypoint-builtin-delve` named `entrypoint-builtin`. Is this intended? Maybe rename it in the source instead?

File: test/e2e/docker/Dockerfile.debug [hunk 7]
```diff
@@ -30,6 +31,6 @@ ENV CMTHOME=/cometbft
 ENV GORACE "halt_on_error=1"
 
 EXPOSE 26656 26657 26660 6060 2345 2346
-ENTRYPOINT ["/usr/bin/entrypoint-delve"]
+#ENTRYPOINT ["/usr/bin/entrypoint-delve"]
```
Comment (mzabaluev) [inline test/e2e/docker/Dockerfile.debug:34]: This looks temp-ish, restore or remove the `ENTRYPOINT` directive altogether?

File: consensus/reactor_test.go (hunk 5 above)
Comment (lasarojc) [reply consensus/reactor_test.go:436]: reverted.

File: test/e2e/docker/Dockerfile.debug (hunk 6 above)
Comment (lasarojc) [reply test/e2e/docker/Dockerfile.debug]: reverted this change, which was for debugging.
copying was a way to override the entry point.
